cmd/exit_handler.go

```diff
@@ -0,0 +1,312 @@
+	return &SignalCallback_{
+		callback,
+		priority,
+	}
+}
+
+// GetPriority is a getter for callback's priority
+func (s *SignalCallback_) GetPriority() Priority {
+	return s.priority
+}
+
+// Call executes callback
+func (s *SignalCallback_) Call() {
+	s.callbackFunc()
+}
+
+// SignalHandler sends Signal to listeners and call registered callbacks
+type SignalHandler_ struct {
+	ch        chan os.Signal
+	listeners []net.Listener
+	callbacks []*SignalCallback_
+	signals   []os.Signal
+}
+
+// NewSignalHandler returns new SignalHandler registered for particular os.Signals
+func NewSignalHandler_(handledSignals []os.Signal) (*SignalHandler_, error) {
+	return &SignalHandler_{ch: make(chan os.Signal), signals: handledSignals}, nil
+}
+
+// AddListener to listeners list
+func (handler *SignalHandler_) AddListener(listener net.Listener) {
+	handler.listeners = append(handler.listeners, listener)
+}
+
+// GetChannel returns channel of os.Signal
+func (handler *SignalHandler_) GetChannel() chan os.Signal {
+	return handler.ch
+}
+
+// AddCallback to callbacks list
+func (handler *SignalHandler_) AddCallback(callback *SignalCallback_) {
+	handler.callbacks = append(handler.callbacks, callback)
+}
+
+// Register should be called as goroutine
+func (handler *SignalHandler_) Register() {
+	signal.Notify(handler.ch, handler.signals...)
+
+	<-handler.ch
+
+	for _, listener := range handler.listeners {
+		listener.Close()
+	}
+	for _, callback := range handler.callbacks {
+		callback.Call()
+	}
+	os.Exit(0)
+}
+
+// AcraExitHandler is a common interface for exit handlers of Acra services
+type AcraExitHandler interface {
+	AddExitSignalHandler(signal AcraSignal, priority Priority)
+	AddDeferFunc(deferFunc func(), priority Priority)
+	ExitZero()
+	ExitOne()
+}
+
+// ExitHandler is an implementation of AcraExitHandler
+type ExitHandler struct {
+	deferFunctions []DeferFunction
+	signalHandlers []AcraSignal
+	notification   chan os.Signal
+}
+
+// NewExitHandler is a constructor for ExitHandler
+func NewExitHandler() (*ExitHandler, error) {
+	return &ExitHandler{
+		deferFunctions: nil,
+		signalHandlers: nil,
+		notification:   make(chan os.Signal),
+	}, nil
+}
+
+// NewSigTERMHandler is a constructor for SigTERMHandler
+func NewSigTERMHandler() AcraSignal {
+	return &SigTERMHandler{
+		nil,
+		nil,
+	}
+}
+
+// NewSigINTHandler is a constructor for SigINTHandler
+func NewSigINTHandler() AcraSignal {
+	return &SigINTHandler{
+		NewSigTERMHandler().(*SigTERMHandler),
+	}
+}
+
+// AddExitSignalHandler appends new signal for handling
+func (s *ExitHandler) AddExitSignalHandler(input AcraSignal) {
+	s.signalHandlers = append(s.signalHandlers, input)
+}
+
+func (s *ExitHandler) waitForExitSystemSignal(exitFuncOnSigINT func(), exitFuncOnSigTERM func()) {
+	signal.Notify(s.notification, s.getSignals()...)
+
+	switch <-s.notification {
+	case syscall.SIGINT:
+		exitFuncOnSigINT()
+	case syscall.SIGTERM:
+		exitFuncOnSigTERM()
+	}
```

Comment:
This seems to be kinda hardcoded. Will be add more method arguments for each signal?

For example, the signal handler could return the exit code to use (if the signal handler wants the process to exit after the handling is complete).